cmd/setec/setec.go

```diff
@@ -35,97 +38,90 @@ import (
+The first time you run the server, you must provide a TS_AUTHKEY to authorize
+the node on the tailnet.
+
+With the --dev flag, the server runs with a dummy KMS. This mode is intended
+for debugging and is NOT SAFE for production use.
+
+Otherwise you must provide a --kms-key-name to use to encrypt the database.`,
+
+				SetFlags: func(_ *command.Env, fs *flag.FlagSet) { flax.MustBind(fs, &serverArgs) },
+				Run:      runServer,
+			},
+			{
+				Name:  "list",
+				Usage: "[flags] <server>",
+				Help:  "List all secrets visible to the caller.",
+				Run:   runList,
+			},
+			{
+				Name:  "info",
+				Usage: "<server> <secret-name>",
+				Help:  "Get metadata for the specified secret.",
+				Run:   runInfo,
+			},
+			{
+				Name:  "get",
+				Usage: "[flags] <server> <secret-name>",
+				Help: `Get the active value of the specified secret.
+
+With --version, fetch the specified version instead of the active one.
+With --if-changed, return the active value only if it differs from --version.`,
+
+				SetFlags: func(_ *command.Env, fs *flag.FlagSet) { flax.MustBind(fs, &getArgs) },
+				Run:      runGet,
+			},
+			{
+				Name:  "put",
+				Usage: "setec [flags] <server> <secret-name>",
+				Help: `Put a new value for the specified secret.
+
+With --from-file, the new value is read from the specified file; otherwise
+the user is prompted for a new value and confirmation at the terminal.`,
+
+				SetFlags: func(_ *command.Env, fs *flag.FlagSet) { flax.MustBind(fs, &putArgs) },
+				Run:      runPut,
+			},
+			{
+				Name:  "set-active",
+				Usage: "<server> <secret-name> <secret-version>",
+				Help:  "Set the active version of the specified secret.",
+				Run:   runSetActive,
+			},
+			command.HelpCommand(nil),
 		},
-		Exec: func(context.Context, []string) error { return flag.ErrHelp },
-	}
-
-	if err := root.ParseAndRun(context.Background(), os.Args[1:]); errors.Is(err, flag.ErrHelp) {
-		// A command tried to run but instructed us to print
-		// help. Exit unsuccessfully to convey that the intended
-		// command didn't run. Note, this branch isn't taken when the
-		// user requests --help explicitly.
-		os.Exit(2)
-	} else if err != nil {
-		log.Fatal(err)
 	}
+	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
+	defer cancel()
+	command.RunOrFail(root.NewEnv(nil).SetContext(ctx), os.Args[1:])
 }
 
 var serverArgs struct {
-	StateDir   string
-	Hostname   string
-	KMSKeyName string
-	Dev        bool
+	StateDir   string `flag:"state-dir,Server state directory"`
+	Hostname   string `flag:"hostname,Tailscale hostname to use"`
+	KMSKeyName string `flag:"kms-key-name,Name of KMS key to use for database encryption"`
+	Dev        bool   `flag:"dev,Run in developer mode"`
 }
 
-func runServer(ctx context.Context, args []string) error {
-	if len(args) > 0 {
-		return flag.ErrHelp
+func runServer(env *command.Env) error {
+	if len(env.Args) > 0 {
```

Comment:
Yeah, I'm happy to iterate on that some more.

I've thought about a few different ways to handle that, but all of them get fiddly as soon as you have something like "I want two or three, either <a> <b> <c> or <a> <c>" which is not uncommon, although we don't use it here.

Reply:
The "index as flag name" is really cute, and I enjoy how it fits into the model, but I worry it might get a little hard to administer in non-obvious cases.

Reply:
Thanks, this was a good inspiration. I re-worked it using the adapters we discussed, now implemented.